blockchain-watcher/src/domain/actions/RunDeploymentContract.ts

```diff
@@ -0,0 +1,96 @@
+import { ConsumerMessageRepository, StatRepository } from "../repositories";
+import { DeploymentContractConfig } from "../deploymentContract/DeploymentContract";
+import { setTimeout } from "timers/promises";
+import { Handler } from "../entities";
+import winston from "winston";
+
+const PREFIX = "deployment_contract";
+const ID = "deployment-contract";
+
+export abstract class RunDeploymentContract {
+  private consumerMessageRepository: ConsumerMessageRepository;
+  private statRepo?: StatRepository;
+  private interval: number = 1_000;
+  private running: boolean = false;
+  private id: string;
+
+  protected abstract logger: winston.Logger;
+  protected abstract execute(message: any): Promise<any>;
+  protected abstract report(): void;
+  protected abstract save(message: any[]): Promise<any>;
+
+  constructor(
+    statsRepo: StatRepository,
+    consumerMessageRepository: ConsumerMessageRepository,
+    cfg: DeploymentContractConfig
+  ) {
+    this.consumerMessageRepository = consumerMessageRepository;
+    this.interval = cfg.interval;
+    this.statRepo = statsRepo;
+    this.running = true;
+    this.id = cfg.id;
+  }
+
+  public async run(): Promise<void> {
+    while (this.running) {
+      const jobStartTime = performance.now();
+
+      try {
+        this.report();
+
+        // Get messages from the queue
+        const messages = await this.consumerMessageRepository.handleMessage();
+        this.logger.info(`[run] Processing ${messages.length} messages`);
+
+        const messagesToProcess = await this.executeProcess(messages);
+        if (messagesToProcess.length < 0) return undefined;
+
+        // Save messages to the database
+        await this.save(messagesToProcess);
+
+        // Delete messages from the queue
+        await this.deleteMessages(messages);
+
+        const jobEndTime = performance.now();
+        const jobExecutionTime = Number(((jobEndTime - jobStartTime) / 1000).toFixed(2));
+
+        this.statRepo?.measure(`${PREFIX}_execution_time`, jobExecutionTime, { job: this.id });
+        this.statRepo?.count(`${PREFIX}_items_total`, { id: this.id }, messages.length);
+      } catch (e: Error | any) {
+        this.logger.error("[run] Error processing messages", e);
+        this.statRepo?.count(`${PREFIX}_runs_total`, { id: ID, status: "error" });
+        await setTimeout(this.interval);
+      }
+
+      this.statRepo?.count(`${PREFIX}_runs_total`, { id: ID, status: "success" });
+      await setTimeout(this.interval);
+    }
+  }
+
+  private async executeProcess(messages: any[]): Promise<any[]> {
+    try {
+      const messagesToProcess = [];
+      for (const message of messages) {
+        const result = await this.execute(message);
+        if (result) {
+          messagesToProcess.push(message);
+        }
+      }
+      return messagesToProcess;
```

Comment:
If there is no reason to fear concurrency issues you can do:
```suggestion
      const results = await Promise.all(messages.map(m => this.execute(m));
      return results.filter(r => !!r);
```